connman: 1.37 -> 1.38

doronbehar opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

- Update connman 1.38, see [change log](https://git.kernel.org/pub/scm/network/connman/connman.git/commit/?id=1ee420ace2b8edb0d4025f469aaa3d00d220dc98).
- Cleanup build for easy enable / disable of features.

###### Things done

- [X] Rewrite all features' enable / disable flags.
- [X] Cleaned up dependencies, including unneeded ones.
- [X] Added 2 new build flavors - `connmanFull` and `connmanThin` with several more / less features enabled / disabled.
- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @matejc

doronbehar committed:
connman: format argument lists and inputs

doronbehar committed:
connman: 1.37 -> 1.38

doronbehar committed:
connman: remove unneeded deps

doronbehar committed:
connman: remove unneeded hooks

doronbehar committed:
connman: quote out in configure flag

doronbehar committed:
connman: quote homepage

doronbehar committed:
connman: upgrade to readline latest

doronbehar committed:
connman: doCheck and split outputs

doronbehar committed:
connman: cleanup build from the groundup

doronbehar committed:
connman: Add 2 more build flavors

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/122

erikarvstedt commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2464,6 +2464,36 @@ in
   conspy = callPackage ../os-specific/linux/conspy {};
 
   connman = callPackage ../tools/networking/connman { };
+  connmanThin = callPackage ../tools/networking/connman {
```
`connmanMinimal` is more idiomatic.

These long definitions belong to a separate file. Use the definition of `tesseract` as a template:
```
  inherit (callPackage ../tools/networking/connman {})
    connman
    connmanMinimal
    connmanFull;
```

doronbehar replied:
I'm not sure I understand the idea of this syntax.. Where should I put all of the `enableThis = true/false;` flags?

doronbehar replied:
Got it.

erikarvstedt commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2464,6 +2464,36 @@ in
+    enablePolkit = false;
+    enablePptp = false;
+    enableLoopback = false;
+    # enableEthernet = false;
```
Add a comment why these are uncommented.

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -1,59 +1,171 @@
+, libmnl
+, gnutls
+, readline
+# Choices one has to decide
```
```suggestion
# configurable options
```

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -1,59 +1,171 @@
+  ;
 
+  # Fix file program not found
   preConfigure = ''
```
```suggestion
  patchPhase = ''
```

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -1,59 +1,171 @@
-    "--enable-bluetooth"
-    "--enable-wifi"
-    "--enable-polkit"
+    # This is for building and running tests (probably enabled by default),
```
```suggestion
# for building and running tests
# --enable-tests installs the tests, we don't want that
"--enable-tools"
```

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -1,59 +1,171 @@
+    ++ stdenv.lib.optionals (!enableWifi) [ "--disable-wifi" ]
+    # We (almost) always turn on IWD support as it doesn't require any new dependencies
+    # and it's easier for the NixOS module to use only 1 connmand package when
+    # IWD is requested
```
```suggestion
    # enable IWD support for wifi as it doesn't require any new dependencies
    # and it's easier for the NixOS module to use only one connman package when
    # IWD is requested
```

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -1,59 +1,171 @@
+    ++ stdenv.lib.optionals (firewallType == "nftables") [ libnftnl ]
+  ;
 
+  # Fix file program not found
```
```suggestion
  # fix invalid path to 'file'
```

erikarvstedt commented:
Nice overall improvements, thanks! :smile_cat:

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -1,59 +1,171 @@
-    "--enable-wifi"
-    "--enable-polkit"
+    # This is for building and running tests (probably enabled by default),
+    # --enable-tests installs the tests as well
```
```suggestion
    # for building and running tests
    # --enable-tests installs the tests, we don't want that
```

doronbehar committed:
connman: Fix comments after @erikarvstedt's comments

doronbehar committed:
connman: Use global reference to lib.optionals

doronbehar committed:
connman: Use patchPhase instead of preConfigure

doronbehar reviewed (commented):
Thanks for reviewing. I have just 1 doubt left.

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -52,6 +52,8 @@
 assert stdenv.lib.asserts.assertOneOf "dnsType" dnsType [ "internal" "systemd-resolved" ];
 
+let inherit (stdenv.lib) optionals; 
+
```
This is currently broken
```suggestion
let 
  inherit (stdenv.lib) optionals;
in
```

doronbehar replied:
Right, I haven't tested that. I got it fixed in last push.

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -84,8 +84,10 @@ stdenv.mkDerivation rec {
 
   # fix invalid path to 'file'
-  preConfigure = ''
+  patchPhase = ''
```
Sorry, I actually meant to suggest using `postPatch`,  so you don't have to run the hooks manually.

doronbehar replied:
Fixed.

erikarvstedt commented:
> I'm not sure I understand the idea of this syntax
Just copy the approach that tesseract uses. Rename `default.nix` to `connman.nix` and move the definitions of the three `connman` variants into default.nix

erikarvstedt commented on pkgs/tools/networking/connman/default.nix:
```diff
@@ -99,17 +99,17 @@ stdenv.mkDerivation rec {
-    # This is for building and running tests (probably enabled by default),
-    # --enable-tests installs the tests as well
+    # for building and running tests
+    # --enable-tests # installs the tests, we don't want that
```
Sorry, that single `#` slipped in there, you can remove it again.
```suggestion
    # --enable-tests installs the tests, we don't want that
```

erikarvstedt replied:
Ok, that's fine.

doronbehar committed:
connman: declare build flavors idiomaticly

Use tesseract as a source for inspiration - rename default.nix ->
connman.nix and declare build flavors in default.nix .

doronbehar committed:
connman: Switch to postPatch from patchPhase

doronbehar commented:
Great, thanks again @erikarvstedt . PR should be ready now.

erikarvstedt commented:
I think that all options that just translate to a `--enable/disable-*` configure flag, without additional logic like setting `buildInputs`, can be removed. It's not worth the added complexity, the user can always override `configureFlags`.

Here's how it [should be done](https://github.com/erikarvstedt/nixpkgs/commit/1860c678353d1a9d9706bff7581abcb4a5cc0944). I've also added these fixes:
- moved runtime deps to `buildInputs`
- removed use of `lowPrio`

What's left to be done: 
- I can't judge if the profiles `connmanFull` and `connmanMinimal` are sensible. This needs feedback from other connman users.
- The [AUR build](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=connman-git) supports these optional runtime dependencies: `bluez`, `wpa_supplicant`, `iwd`. Maybe we should provide these, too.
Although we enable `iwd`, we don't have `iwd` in `buildInputs`. Is that correct?

doronbehar commented:
> I think that all options that just translate to a --enable/disable-* configure flag, without additional logic like setting buildInputs, can be removed. It's not worth the added complexity, the user can always override configureFlags.

I'm not sure I agree. The reason behind my approach is that the --disable/ --enable flags are more explicit - we either require a certain feature or we don't, meaning we don't count on it's presence in `buildInputs` or a like to provide it's compilation. I got convinced that this is a good approach after: https://github.com/NixOS/nixpkgs/pull/79864#issuecomment-591089959.

In the current state of `configureFlags`, we use the same defaults as upstream, only it's easier to figure out which feature is enabled / disabled by default. Perhaps the only exception is `--enable-iwd` (which is disabled by default in upstream's `./configure`). This way, you don't have to download the tarball yourself and run `./configure --help` in order to view all available options and see which are enabled/disabled by default and _then_, compare it with what's in `connman.nix`.

> - removed use of lowPrio

I don't mind this change though :)

> I can't judge if the profiles connmanFull and connmanMinimal and all of the added package options are sensible. This needs feedback from other connman users.

I agree. I am a connman user myself and I can imagine myself using at least `connmanMinimal` - perhaps on an embedded device that doesn't need WiFi nor any other feature. As for `connmanFull`, I think there would always be someone out there who'd want all the bells and whistles this program can provide. That's just me though. I'm open to hear other opinions.

> Although we enable iwd, we don't have iwd in buildInputs. Is that correct?

Correct. The same goes for wpa_supplicant. It's evident that connman (in 1.37 as well) connects to both daemons via dbus. Hence, the user should choose which program he wishes to use as a wireless daemon and install it. Even Arch Linux puts them in `makedepends` and not `depends`.

erikarvstedt commented:
Thanks for your thoughtful response.

> I'm not sure I agree.

Yeah, we'll have to wait for a judgement from a more seasoned contributor. 
Can you add my other fixes that are unrelated to the options?

doronbehar commented:
> Yeah, we'll have to wait for a judgement from a more seasoned contributor.

Agreed. Allow me to summon @worldofpeace to judge on this issue as he (they) supported this approach in https://github.com/NixOS/nixpkgs/pull/79864#issuecomment-591089959 

> Can you add my other fixes that are unrelated to the options?

Sure. As for the removal of `lowPrio`, you got it. But as for:

> - moved runtime deps to `buildInputs`

I see that in https://github.com/erikarvstedt/nixpkgs/blob/1860c678353d1a9d9706bff7581abcb4a5cc0944/pkgs/tools/networking/connman/connman.nix#L47 . My opinion is that this is an imposition we shouldn't take. Just because a connman derivation was built with (e.g) a certain vpn feature enabled, it doesn't mean that  all of the (e.g) openvpn derivation should be downloaded to a user's `/nix/store`. If a user would like to couple connman with say openvpn, he should handle that in `/etc/nixos/configuration.nix` somehow and not most people wouldn't do that.

erikarvstedt commented:
The distinction between `buildInputs` and `nativeBuildinputs` is only relevant for cross-compiling. In regular builds (no cross-compilation), both can be used interchangeably and won't affect which drvs are dependencies (`nix-store -q --references`) of the resulting build output.

Actually, it seems that connman just references the openvpn binary (`sbin/openvpn`) in `connman-vpnd`. A dependency should only be added to `*buildInputs` when it should modify the build environment in some way, like setting up library paths. This is not needed for `openvpn` and in fact most of the other `buildInputs`, so they can be removed. Here's my patch: https://github.com/erikarvstedt/nixpkgs/commit/cd1273bca9275ea6bb6a5d0ea440521db34d66d2.

doronbehar commented:
I always thought that the difference between `buildInputs` and `nativeBuildInputs` is as between runtime dependencies and build only dependencies (`makedepends`). I'm using NixOS for a long time and I think I got finally a little bit better understanding of I dependencies in Nixpkgs.. I still get lost though when reading https://nixos.org/nixpkgs/manual/#ssec-stdenv-dependencies .

@erikarvstedt I know you won't completely agree with me about this but I'd like to ask you anyway perhaps only for the sake of me learning: Say I have several binaries in `./result/bin/*`. I can run:

```
$ ldd ./result/bin/*
```

<details><summary>Which gives:</summary>

```
./result/bin/connmanctl:
	linux-vdso.so.1 (0x00007ffffdd1a000)
	libdbus-1.so.3 => /nix/store/srvy3fwc6655kq9a274dfaw8iczna9b9-dbus-1.12.16-lib/lib/libdbus-1.so.3 (0x00007fb48c3e5000)
	libglib-2.0.so.0 => /nix/store/vwvjrh97byihk40gwl26f0lnb3w405cv-glib-2.62.4/lib/libglib-2.0.so.0 (0x00007fb48c2be000)
	libreadline.so.6 => /nix/store/j0hlpbcs6j3phqxsv09wjrvwxv17lg50-readline-6.3p08/lib/libreadline.so.6 (0x00007fb48c274000)
	libdl.so.2 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libdl.so.2 (0x00007fb48c26f000)
	libc.so.6 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libc.so.6 (0x00007fb48c0b0000)
	libpthread.so.0 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libpthread.so.0 (0x00007fb48c08d000)
	libsystemd.so.0 => /nix/store/2n66ha7fnk87jr0yq6ihnl8czwa7iknw-systemd-243.4-lib/lib/libsystemd.so.0 (0x00007fb48bfd1000)
	libpcre.so.1 => /nix/store/b5sam900j8278388zwirpwap5gw8m4l4-pcre-8.43/lib/libpcre.so.1 (0x00007fb48bf5e000)
	libncursesw.so.6 => /nix/store/7yjv8l75a8b1x3vmmnbsdwd5b0gzngga-ncurses-6.1-20190112/lib/libncursesw.so.6 (0x00007fb48beed000)
	/nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/ld-linux-x86-64.so.2 => /nix/store/y9zg6ryffgc5c9y67fcmfdkyyiivjzpj-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007fb48c43b000)
	librt.so.1 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/librt.so.1 (0x00007fb48bee1000)
	liblzma.so.5 => /nix/store/yvsjlr772hagwpdwg9qlsfpdf68wyjmg-xz-5.2.4/lib/liblzma.so.5 (0x00007fb48beb7000)
	liblz4.so.1 => /nix/store/5handm3cwbix8zys1llznzdcxpxwj06i-lz4-1.9.1/lib/liblz4.so.1 (0x00007fb48be84000)
	libcap.so.2 => /nix/store/3yc4gjawd1w9gvp1539li3lil2icdzdz-libcap-2.27-lib/lib/libcap.so.2 (0x00007fb48be7d000)
	libgcrypt.so.20 => /nix/store/7wjbixabz2vq6kp0wfp64djg6bcb0x5r-libgcrypt-1.8.5/lib/libgcrypt.so.20 (0x00007fb48bd5e000)
	libgpg-error.so.0 => /nix/store/zzipnaayyzfvga0q3c70y861d58gdjc0-libgpg-error-1.36/lib/libgpg-error.so.0 (0x00007fb48bd38000)
./result/bin/connmand:
	linux-vdso.so.1 (0x00007ffe0617c000)
	libglib-2.0.so.0 => /nix/store/vwvjrh97byihk40gwl26f0lnb3w405cv-glib-2.62.4/lib/libglib-2.0.so.0 (0x00007f79f65da000)
	libdbus-1.so.3 => /nix/store/srvy3fwc6655kq9a274dfaw8iczna9b9-dbus-1.12.16-lib/lib/libdbus-1.so.3 (0x00007f79f6586000)
	libgnutls.so.30 => /nix/store/5sn08v4z5agv2k659szh95c8nklrcb61-gnutls-3.6.12/lib/libgnutls.so.30 (0x00007f79f63ac000)
	libresolv.so.2 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libresolv.so.2 (0x00007f79f6393000)
	libdl.so.2 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libdl.so.2 (0x00007f79f638e000)
	librt.so.1 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/librt.so.1 (0x00007f79f6382000)
	libxtables.so.12 => /nix/store/7jcgvxhmicwbvl4jw78l4q939f8596ax-iptables-1.8.4/lib/libxtables.so.12 (0x00007f79f636f000)
	libc.so.6 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libc.so.6 (0x00007f79f61b0000)
	libpcre.so.1 => /nix/store/b5sam900j8278388zwirpwap5gw8m4l4-pcre-8.43/lib/libpcre.so.1 (0x00007f79f613d000)
	libpthread.so.0 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libpthread.so.0 (0x00007f79f611c000)
	libsystemd.so.0 => /nix/store/2n66ha7fnk87jr0yq6ihnl8czwa7iknw-systemd-243.4-lib/lib/libsystemd.so.0 (0x00007f79f605e000)
	libp11-kit.so.0 => /nix/store/w02i4w7r4hpycmyryh7xr6aa6l3v1dsb-p11-kit-0.23.19/lib/libp11-kit.so.0 (0x00007f79f5f2d000)
	libunistring.so.2 => /nix/store/mxaxvp33wg9sim8qh2kkw041v492bvxj-libunistring-0.9.10/lib/libunistring.so.2 (0x00007f79f5daa000)
	libtasn1.so.6 => /nix/store/kzmdmkpp960v5d1czbwzzj2zzxrg4wxl-libtasn1-4.16.0/lib/libtasn1.so.6 (0x00007f79f5d94000)
	libnettle.so.7 => /nix/store/g3s0wk5b0d7hcy5xlw7x0fmyfdf5yxy2-nettle-3.5.1/lib/libnettle.so.7 (0x00007f79f5d57000)
	libhogweed.so.5 => /nix/store/g3s0wk5b0d7hcy5xlw7x0fmyfdf5yxy2-nettle-3.5.1/lib/libhogweed.so.5 (0x00007f79f5d1c000)
	libgmp.so.10 => /nix/store/r05vk4zq34sps98izq4v4n9h7m1a6wgx-gmp-6.2.0/lib/libgmp.so.10 (0x00007f79f5c7a000)
	/nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/ld-linux-x86-64.so.2 => /nix/store/y9zg6ryffgc5c9y67fcmfdkyyiivjzpj-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007f79f6703000)
	libpcap.so.1 => /nix/store/1prpyz7vis4wwakqzabca3qpsk6yvb4r-libpcap-1.9.1/lib/libpcap.so.1 (0x00007f79f5c30000)
	libmnl.so.0 => /nix/store/p6mzr73njz01989c685ci346ldmglm5z-libmnl-1.0.4/lib/libmnl.so.0 (0x00007f79f5c28000)
	libnftnl.so.11 => /nix/store/r1f53qxf42mqxfad9sbffibxr0irqdlz-libnftnl-1.1.5/lib/libnftnl.so.11 (0x00007f79f5bf5000)
	liblzma.so.5 => /nix/store/yvsjlr772hagwpdwg9qlsfpdf68wyjmg-xz-5.2.4/lib/liblzma.so.5 (0x00007f79f5bcb000)
	liblz4.so.1 => /nix/store/5handm3cwbix8zys1llznzdcxpxwj06i-lz4-1.9.1/lib/liblz4.so.1 (0x00007f79f5b98000)
	libcap.so.2 => /nix/store/3yc4gjawd1w9gvp1539li3lil2icdzdz-libcap-2.27-lib/lib/libcap.so.2 (0x00007f79f5b91000)
	libgcrypt.so.20 => /nix/store/7wjbixabz2vq6kp0wfp64djg6bcb0x5r-libgcrypt-1.8.5/lib/libgcrypt.so.20 (0x00007f79f5a70000)
	libffi.so.7 => /nix/store/phwyg24nm8z4r91iyrhmdfljxpw1ihd5-libffi-3.3/lib/../lib64/libffi.so.7 (0x00007f79f5a63000)
	libgpg-error.so.0 => /nix/store/zzipnaayyzfvga0q3c70y861d58gdjc0-libgpg-error-1.36/lib/libgpg-error.so.0 (0x00007f79f5a3f000)
./result/bin/connmand-wait-online:
	linux-vdso.so.1 (0x00007ffc367bc000)
	libglib-2.0.so.0 => /nix/store/vwvjrh97byihk40gwl26f0lnb3w405cv-glib-2.62.4/lib/libglib-2.0.so.0 (0x00007f04d9133000)
	libdbus-1.so.3 => /nix/store/srvy3fwc6655kq9a274dfaw8iczna9b9-dbus-1.12.16-lib/lib/libdbus-1.so.3 (0x00007f04d90df000)
	libc.so.6 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libc.so.6 (0x00007f04d8f20000)
	libpcre.so.1 => /nix/store/b5sam900j8278388zwirpwap5gw8m4l4-pcre-8.43/lib/libpcre.so.1 (0x00007f04d8ead000)
	libpthread.so.0 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libpthread.so.0 (0x00007f04d8e8c000)
	libsystemd.so.0 => /nix/store/2n66ha7fnk87jr0yq6ihnl8czwa7iknw-systemd-243.4-lib/lib/libsystemd.so.0 (0x00007f04d8dce000)
	/nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/ld-linux-x86-64.so.2 => /nix/store/y9zg6ryffgc5c9y67fcmfdkyyiivjzpj-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007f04d925c000)
	librt.so.1 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/librt.so.1 (0x00007f04d8dc4000)
	liblzma.so.5 => /nix/store/yvsjlr772hagwpdwg9qlsfpdf68wyjmg-xz-5.2.4/lib/liblzma.so.5 (0x00007f04d8d9a000)
	liblz4.so.1 => /nix/store/5handm3cwbix8zys1llznzdcxpxwj06i-lz4-1.9.1/lib/liblz4.so.1 (0x00007f04d8d67000)
	libcap.so.2 => /nix/store/3yc4gjawd1w9gvp1539li3lil2icdzdz-libcap-2.27-lib/lib/libcap.so.2 (0x00007f04d8d5e000)
	libgcrypt.so.20 => /nix/store/7wjbixabz2vq6kp0wfp64djg6bcb0x5r-libgcrypt-1.8.5/lib/libgcrypt.so.20 (0x00007f04d8c3f000)
	libgpg-error.so.0 => /nix/store/zzipnaayyzfvga0q3c70y861d58gdjc0-libgpg-error-1.36/lib/libgpg-error.so.0 (0x00007f04d8c1b000)
./result/bin/connman-vpnd:
	linux-vdso.so.1 (0x00007ffdff9e8000)
	libglib-2.0.so.0 => /nix/store/vwvjrh97byihk40gwl26f0lnb3w405cv-glib-2.62.4/lib/libglib-2.0.so.0 (0x00007f928de62000)
	libdbus-1.so.3 => /nix/store/srvy3fwc6655kq9a274dfaw8iczna9b9-dbus-1.12.16-lib/lib/libdbus-1.so.3 (0x00007f928de0e000)
	libgnutls.so.30 => /nix/store/5sn08v4z5agv2k659szh95c8nklrcb61-gnutls-3.6.12/lib/libgnutls.so.30 (0x00007f928dc34000)
	libresolv.so.2 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libresolv.so.2 (0x00007f928dc1b000)
	libdl.so.2 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libdl.so.2 (0x00007f928dc16000)
	libc.so.6 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libc.so.6 (0x00007f928da55000)
	libpcre.so.1 => /nix/store/b5sam900j8278388zwirpwap5gw8m4l4-pcre-8.43/lib/libpcre.so.1 (0x00007f928d9e2000)
	libpthread.so.0 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/libpthread.so.0 (0x00007f928d9c1000)
	libsystemd.so.0 => /nix/store/2n66ha7fnk87jr0yq6ihnl8czwa7iknw-systemd-243.4-lib/lib/libsystemd.so.0 (0x00007f928d905000)
	libp11-kit.so.0 => /nix/store/w02i4w7r4hpycmyryh7xr6aa6l3v1dsb-p11-kit-0.23.19/lib/libp11-kit.so.0 (0x00007f928d7d2000)
	libunistring.so.2 => /nix/store/mxaxvp33wg9sim8qh2kkw041v492bvxj-libunistring-0.9.10/lib/libunistring.so.2 (0x00007f928d64f000)
	libtasn1.so.6 => /nix/store/kzmdmkpp960v5d1czbwzzj2zzxrg4wxl-libtasn1-4.16.0/lib/libtasn1.so.6 (0x00007f928d639000)
	libnettle.so.7 => /nix/store/g3s0wk5b0d7hcy5xlw7x0fmyfdf5yxy2-nettle-3.5.1/lib/libnettle.so.7 (0x00007f928d5fe000)
	libhogweed.so.5 => /nix/store/g3s0wk5b0d7hcy5xlw7x0fmyfdf5yxy2-nettle-3.5.1/lib/libhogweed.so.5 (0x00007f928d5c3000)
	libgmp.so.10 => /nix/store/r05vk4zq34sps98izq4v4n9h7m1a6wgx-gmp-6.2.0/lib/libgmp.so.10 (0x00007f928d51f000)
	/nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/ld-linux-x86-64.so.2 => /nix/store/y9zg6ryffgc5c9y67fcmfdkyyiivjzpj-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007f928df8b000)
	librt.so.1 => /nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30/lib/librt.so.1 (0x00007f928d515000)
	liblzma.so.5 => /nix/store/yvsjlr772hagwpdwg9qlsfpdf68wyjmg-xz-5.2.4/lib/liblzma.so.5 (0x00007f928d4eb000)
	liblz4.so.1 => /nix/store/5handm3cwbix8zys1llznzdcxpxwj06i-lz4-1.9.1/lib/liblz4.so.1 (0x00007f928d4b8000)
	libcap.so.2 => /nix/store/3yc4gjawd1w9gvp1539li3lil2icdzdz-libcap-2.27-lib/lib/libcap.so.2 (0x00007f928d4af000)
	libgcrypt.so.20 => /nix/store/7wjbixabz2vq6kp0wfp64djg6bcb0x5r-libgcrypt-1.8.5/lib/libgcrypt.so.20 (0x00007f928d390000)
	libffi.so.7 => /nix/store/phwyg24nm8z4r91iyrhmdfljxpw1ihd5-libffi-3.3/lib/../lib64/libffi.so.7 (0x00007f928d383000)
	libgpg-error.so.0 => /nix/store/zzipnaayyzfvga0q3c70y861d58gdjc0-libgpg-error-1.36/lib/libgpg-error.so.0 (0x00007f928d35f000)
```

</details>
This gives a view of the actual runtime dependencies. As for an executable there calling (e.g) openvpn or a like, we can't ever know that by looking at `ldd`'s output so let's ignore this scenario for the sake of simplicity.

In comparison to `ldd`'s output, this is the output of `nix-store -q --references $(nix-build -A connmanFull)`:

```
/nix/store/1ncwrl8bplq3xhmj8pxfkx4y0i90vmnx-glibc-2.30
/nix/store/00rilbj1qjpr8kzi7rasmsiw9v1czrd6-vpnc-0.5.3-post-r550
/nix/store/5sn08v4z5agv2k659szh95c8nklrcb61-gnutls-3.6.12
/nix/store/p6mzr73njz01989c685ci346ldmglm5z-libmnl-1.0.4
/nix/store/7jcgvxhmicwbvl4jw78l4q939f8596ax-iptables-1.8.4
/nix/store/injg047z5jv8ng0yxa8ds54nfrhhs5g7-ppp-2.4.7
/nix/store/80nkg3kcivbkavb9nxpf7pfd7vhnxvgp-pptp-1.10.0
/nix/store/bmqdvxbsc6zncmbjw8lf82wsmm6jc579-openvpn-2.4.7
/nix/store/j0hlpbcs6j3phqxsv09wjrvwxv17lg50-readline-6.3p08
/nix/store/rvf426ps2bai9fcghdnn0lxc2j7dq1gi-openconnect-8.05
/nix/store/srvy3fwc6655kq9a274dfaw8iczna9b9-dbus-1.12.16-lib
/nix/store/vwvjrh97byihk40gwl26f0lnb3w405cv-glib-2.62.4
/nix/store/wgd0b14rjh59c75rmv35gxma8kghlmxp-connman-1.38
```

Say I would like to remove the reference to ppp, openvpn, vpnc and openconnect. How should I construct the inputs in such a way that they will only be used in build time, but wouldn't be strictly required for running?

doronbehar committed:
connmanMinimal: remove use of lowPrio

erikarvstedt commented:
> How should I construct the inputs in such a way...

connman calls their binaries by their absolute store paths. So, for private use, you could pass wrapper binaries to the build that call the main binary in PATH.

But this is inappropriate for `nixpkgs` where all dependencies should be made explicit.

doronbehar commented:
I see, (I read https://nixos.org/nixos/nix-pills/automatic-runtime-dependencies.html and that helped me understand better).

As for the `buildInputs` vs `nativeBuildInputs`, here's a quote from https://nixos.org/nixpkgs/manual/#ssec-stdenv-dependencies : 

> Since these packages are able to be run at build-time, they are added to the PATH, as described above. But since these packages are only guaranteed to be able to run then, they shouldn't persist as run-time dependencies. This isn't currently enforced, but could be in the future. 

That makes sense with what I understand better only now. However, in our case, the dependencies `openvpn`, `vpnc` and `openconnect` are used in `configureFlags` and they are referenced directly (to their `bin/` path) hence they will always be referenced in the output, no matter that they are added to `nativeBuildInputs` or `buildInputs`. 

I understand now this logic.

erikarvstedt commented on pkgs/tools/networking/connman/connman.nix:
```diff
@@ -21,7 +21,6 @@
 , enableVpnc ? true
 , vpnc ? true
 , enablePolkit ? true
-, polkit ? null
```
Still needed for polkit rules.

doronbehar replied:
What do you mean?

erikarvstedt replied:
`configure.ac` includes these lines:
```
POLKIT_DATADIR="`$PKG_CONFIG --variable=actiondir polkit`"
POLKIT_DATADIR=""
```
The first line accesses polkit data, the second line discards it. So for now it's indeed unused, but that may change unexpectedly. But let's do what AUR does and keep it out.

doronbehar replied:
Fair enough.

erikarvstedt replied:
As I said, it's okay to keep it out, like in AUR. You can remove it again.

doronbehar replied:
I think it's safer to keep it, I don't mind.

erikarvstedt replied:
Okay, agreed.

erikarvstedt commented on pkgs/tools/networking/connman/connman.nix:
```diff
@@ -0,0 +1,177 @@
+, enableDatafiles ? true
+# optional features which are turned *off* by default
+, enableNetworkManager ? false
+, networkmanager ? null
```
Not used.

doronbehar replied:
:+1:.

erikarvstedt commented on pkgs/tools/networking/connman/connman.nix:
```diff
@@ -125,15 +120,15 @@ stdenv.mkDerivation rec {
     ++ optionals (!enableClient) [ "--disable-client" ]
     ++ optionals (!enableDatafiles) [ "--disable-datafiles" ]
     ++ optionals (enableOpenconnect) [
-      "--enable-openconnect=builtin"
```
What's the reason for switching from builtin to plugin?

doronbehar replied:
I haven't found any mention of `=builtin` in the output of `./configure --help`. Hence I figured these configure flags got ignored. I guess this still hasn't bothered the necessary binaries to be used since (e.g) `--with-openconnect=${openconnect}/bin/openconnect` was used as well.

erikarvstedt replied:
See `configure.ac` and `Makefile.plugins`, grep for `OPENCONNECT_BUILTIN`. Without `builtin` these extensions are built a plugins.

doronbehar replied:
:+1: .

erikarvstedt commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2463,37 +2463,11 @@ in
-    enableTist = true;
-  };
+  inherit (callPackage ../tools/networking/connman {})
+    connmanFull
```
Please reverse the order. Same in `connman.nix`.

doronbehar replied:
:+1:.

doronbehar committed:
connman: remove import of networkmanager

doronbehar committed:
connman: reverse order of build flavors declaration

doronbehar committed:
connman: remove unneeded deps from nativeBuildInputs

alyssais commented:
> Allow me to summon @worldofpeace to judge on this issue as he (they) supported this approach in [#79864 (comment)](https://github.com/NixOS/nixpkgs/pull/79864#issuecomment-591089959)

We do not address worldofpeace like this. I understand they have already made it plain to you that it is not appropriate to call them “he”, including “he (they)”. In case this is still not clear, a phrasing that is not disrespectful to worldofpeace would have been:

“Allow me to summon @worldofpeace to judge on this issue as they supported this approach in…”

erikarvstedt commented:
@alyssais, thanks for reading through all of this. Do you think it would make sense to close this PR and reopen it with squashed commits and one single comment by me which addresses outstanding issues, so we can spare future reviewers all the detours in our discussion?

doronbehar commented:
> We do not address worldofpeace like this. I understand they have already made it plain to you that it is not appropriate to call them “he”, including “he (they)”.

Thank you @alyssais for joining the discussion. Unfortunately, I've failed to understand that a while ago when worldofpeace have first explained it to me and I understood a totally different scenario. Just a few hours ago I've messaged them explaining this. Please forgive my insensitivity.

worldofpeace commented:
> @alyssais, thanks for reading through all of this. Do you think it would make sense to close this PR and reopen it with squashed commits and one single comment by me which addresses outstanding issues, so we can spare future reviewers all the detours in our discussion?

I think it would make sense to re-pr so it can attract reviewers. I currently won't have the time to review this just yet, since we are so close to getting 20.03 out.

doronbehar commented:
Closing for the sake of a cleaner re-pr.
